src/style-spec/expression/definitions/index.js

```diff
@@ -339,32 +341,32 @@ CompoundExpression.register(expressions, {
     '&&': [
         BooleanType,
         varargs(BooleanType),
-        (...args) => args.reduce((a, b) => a && b, true)
+        (ctx, args) => args.reduce((a, b) => a && b.evaluate(ctx), true)
```

Comment:
With evaluation now happening within these definitions, `&&` and `||`  could be implemented so as not to evaluate all their arguments, without needing to be defined in a separate class.   I'm in favor of doing that -- it's true that `CompoundExpression` would then include both normal and special forms, but is there any advantage to avoiding that?

Reply:
No big advantage that I see. This change itself does fix short circuiting, though we should still add tests for that.

Reply:
> This change itself does fix short circuiting

Oy -- yes, that it does 🤦‍♂️